src/Symfony/Component/Messenger/Transport/Schedule/Progress/Progress.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport\Schedule\Progress;
+
+use Symfony\Component\Clock\ClockInterface;
+use Symfony\Component\Lock\LockInterface;
+use Symfony\Contracts\Cache\CacheInterface;
+
+final class Progress implements ProgressInterface
+{
+    private ?\DateTimeImmutable $nextTime;
+    private \DateTimeImmutable $time;
+    private int $index;
+
+    public function __construct(
+        private readonly ClockInterface $clock,
+        private readonly LockInterface $lock,
+        private readonly CacheInterface $cache,
+        private readonly string $name,
+    ) {
+        $this->nextTime = new \DateTimeImmutable('@0');
+        $this->loadProgress($clock->now());
+    }
+
+    public function acquire(): ?\DateTimeImmutable
+    {
+        if (!$this->nextTime || !$this->lock->acquire()) {
+            return null;
+        }
+
+        $now = $this->clock->now();
+        if ($now < $this->nextTime) {
+            return null;
+        }
+
+        [$this->time, $this->index] = $this->loadProgress($now);
+
+        return $now;
+    }
+
+    public function time(): \DateTimeImmutable
+    {
+        return $this->time;
+    }
+
+    public function index(): int
+    {
+        return $this->index;
+    }
+
+    public function save(\DateTimeImmutable $time, int $index): void
+    {
+        $this->time = $time;
+        $this->index = $index;
+
+        $this->cache->get($this->name, fn () => [$time, $index], \INF);
+        $this->lock->refresh();
+    }
+
+    public function release(?\DateTimeImmutable $nextTime): void
+    {
+        $this->nextTime = $nextTime;
+
+        unset($this->time, $this->index);
```

Comment:
We should probably set them to `null` instead.

Reply:
Probably...

The `time` and the `index` properties are not nullable, but must be undefined before `acquire` and after `release`.

Reply:
I've added exceptions. I hope they help to make logic clear.